src/main/java/io/github/in_toto/models/Link.java

```diff
@@ -35,8 +35,8 @@
      */
     public Link(HashMap<String, ArtifactHash> materials,
             HashMap<String, ArtifactHash> products, String name,
-            HashMap<String, String> environment, ArrayList<String> command,
-            HashMap<String, String> byproducts) {
+            HashMap<String, Object> environment, ArrayList<String> command,
+            HashMap<String, Object> byproducts) {
```

Comment:
I'm unsure about these `String` to `Object`changes. Aren't we assuming that the `key:value` pairs are Strings? (you do claim that in the canonicalize function itself.

Reply:
I actually only claim that it's okay to assume that the [`HashMap` **keys** are strings](https://github.com/in-toto/in-toto-java/pull/12/files/d84ecf62ad3ab087add0cdf3c3a92c93a9444935#diff-716698852f0323b14edbffd178b65cacR86).

Here is what the specs say about [byproducts](https://github.com/in-toto/docs/blame/master/in-toto-spec.md#L927) and [environment](https://github.com/in-toto/docs/blame/master/in-toto-spec.md#L941):
>The `"byproducts"` field is an opaque dictionary
...
The format of the environment information is not mandated by the in-toto
specification

The metadata examples, on the other hand, show e.g. [`byproducts.return-value` as `string`](https://github.com/in-toto/docs/blame/v0.9/in-toto-spec.md#L1134), and [`environment.variables` as `list`](https://github.com/in-toto/docs/blame/master/in-toto-spec.md#L958-L965).

In order to provide cross-implementation compatibility (for signatures), we need to decide on a strict metadata format. We can either do this in the specification, or just internally for our different implementations. FWIW, in the reference implementation I currently let the json parser decide about the datatypes and that's also what I tried to do here.

What do you think?

Reply:
I think this is ok. It's probably messy on the jenkins plugin side, but I imagine that'd require correcting on there...